### ISSUE TITLE
ContravariantRegistrationSource will create registrations with illegal ValueType conversions.

### ISSUE THREAD

**Issue by dammejed** [autofac/Autofac#790]
According to MSDN, variance does not apply when the type parameter is a value type, but the ContravariantRegistrationSource will happily attempt to do so anyway.

A somewhat contrived example: 
```csharp
var builder = new ContainerBuilder();
builder.RegisterSource(new ContravariantRegistrationSource());
builder.RegisterType<ObjectHandler>().As<IHandler<object>>();
var container = builder.Build();
//This blows up with an InvalidCastException instead of the expected ComponentNotRegisteredException.
var handler = container.Resolve<IHandler<FooEnum>>();
```

I don't expect this would ever be encountered as written here, but I was encountering some strange interactions between the ContravariantRegistrationSource and RegisterGeneric, related to #341. As that issue is much more difficult to address in a rigorous way, seems better to try to address this isolated case.

**Comment by tillig** [comment]
You're right - [I found the MSDN reference](https://msdn.microsoft.com/en-us/library/dd799517(v=vs.110).aspx).

I'll grab that PR. Thanks!

**Comment by tillig** [comment]
Grabbed the PR and merged. Thanks again!

**Comment by tillig** [comment]
Fix included in v4.1.1, pushed to NuGet.

**Comment by dammejed** [comment]
Thank you!
